java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchTreeBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,110 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import java.util.TreeSet;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.vector.BaseFixedWidthVector;
+import org.apache.arrow.vector.BaseVariableWidthVector;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * The dictionary based on a search tree.
+ * So each add operation can be finished in O(log(n)),
+ * where n is the current dictionary size.
+ * @param <V> the dictionary vector type.
+ */
+public class SearchTreeBasedDictionaryBuilder<V extends ValueVector> extends DictionaryBuilder<V> {
+
+  private final BaseFixedWidthVector fixedWidthVector;
+
+  private final BaseVariableWidthVector variableWidthVector;
+
+  /**
+   * The search tree for storing the value index.
+   */
+  protected TreeSet<Integer> searchTree =
+          new TreeSet<>((index1, index2) -> comparator.compare(index1, index2));
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   * @param encodeNull if null should be encoded.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator, boolean encodeNull) {
+    super(dictionary, comparator, encodeNull);
+
+    if (dictionary instanceof BaseFixedWidthVector) {
+      fixedWidthVector = (BaseFixedWidthVector) dictionary;
+      variableWidthVector = null;
+    } else if (dictionary instanceof BaseVariableWidthVector) {
+      fixedWidthVector = null;
+      variableWidthVector = (BaseVariableWidthVector) dictionary;
+    } else {
+      throw new IllegalArgumentException("The dictionary must be a BaseFixedWidthVector or a BaseVariableWidthVector");
+    }
+  }
+
+  @Override
+  public boolean addValue(V targetVector, int targetIndex) {
+    totalValueCount += 1;
+
+    // first copy the value to the end of the dictionary
+    if (fixedWidthVector != null) {
+      fixedWidthVector.copyFromSafe(targetIndex, distinctValueCount, (BaseFixedWidthVector) targetVector);
```

Comment:
is it possible to move copyFromSafe up to a common interface?  it seems like the only reason why we check variable width vs fixedwidth  is to cast appropriately?

Reply:
Good suggestion. Moved the methods to the ValueVector interface.